File: LeetCode.com/36-ValidSudoku.cs
//36. Valid Sudoku

public class Solution {
    public bool IsValidSudoku(char[][] board) {
        // iterate through rows, then columns
            for (int i = 0; i < 9; i++)
            {
                //is valid columns 1-9, every row
                if (!isParticallyValid(board, i, 0, i, 8))
                    return false;

                //is valid rows 1-9, every column
                if (!isParticallyValid(board, 0, i, 8, i))
                    return false;
            }

            //iterate through subboxes
            for (int i = 0; i < 3; i++)
            {
                for (int j = 0; j < 3; j++)
                {
                    if (!isParticallyValid(board, i * 3, j * 3, i * 3 + 2, j * 3 + 2))
                        return false;
                }
            }
        
        return true;
    }
    
    
    public bool isParticallyValid(char[][] board, int x1, int y1, int x2, int y2)
        {
            var set = new HashSet<char>(); //do not have duplicates
            for (int row = x1; row <= x2; row++)
            {
                for (int col = y1; col <= y2; col++)
                {
                    if (board[row][col] != '.') //ignore blanks
                    {
                        if (!set.Add(board[row][col])) //if not added, invalid board
                            return false;
                    }
                }
            }

            return true;
        }
    
}

File: LeetCode Java/Data Structure I/36-ValidSudoku.java
/*
36. Valid Sudoku
Determine if a 9 x 9 Sudoku board is valid. Only the filled cells need to be validated according to the following rules:
    Each row must contain the digits 1-9 without repetition.
    Each column must contain the digits 1-9 without repetition.
    Each of the nine 3 x 3 sub-boxes of the grid must contain the digits 1-9 without repetition.
Note:
    A Sudoku board (partially filled) could be valid but is not necessarily solvable.
    Only the filled cells need to be validated according to the mentioned rules.
*/
class Solution {
    //O(n^2) space complexity
    public boolean isValidSudoku(char[][] board) {
        HashSet<Character> currRow = new HashSet<Character>();
        HashSet<Character> currCol = new HashSet<Character>();
        HashSet<Character> currBox = new HashSet<Character>(); 
        for (int r = 0; r<board.length; r++) {
            for (int c = 0; c<board.length; c++) {
                if (board[r][c] != '.') { //rows
                    if (!currRow.contains(board[r][c])) {
                        currRow.add(board[r][c]);
                    } else { return false; }
                }
                //columns - just flip r and c
                if (board[c][r] != '.') {
                    if (!currCol.contains(board[c][r])) {
                        currCol.add(board[c][r]);
                    } else { return false; }
                }
                //will only occur once for each box (top left of each box)
                if (r % 3 == 0 && c % 3 == 0) { 
                    currBox.removeAll(currBox);

                    for (int i = 0; i < 3; i++) {
                        for (int k = 0; k < 3; k++) {
                            if (board[i+r][k+c] != '.') {
                                if (!currBox.contains(board[i+r][k+c])) {
                                    currBox.add(board[i+r][k+c]);
                                } else {
                                    return false;
                                }
                            }   
                        }
                    }
                }
            }
            currCol.removeAll(currCol);
            currRow.removeAll(currRow);
        }
        return true;
    }
}